HeapTracker/README.md

```diff
@@ -0,0 +1,128 @@
+    getFreeHeapMemory_malloc --> Heap status: max available(143360 bytes), allocated (0 bytes)
+    Heap Tracker lib: malloc(1024)... INFO: available heap (142336 bytes)
+    Heap Tracker lib: _free(0xbeeebb40, 1024)... INFO: available heap (143360 bytes)
+    Heap Tracker lib: malloc(2048)... INFO: available heap (141312 bytes)
+    Heap Tracker lib: _free(0xbee84640, 2048)... INFO: available heap (143360 bytes)
+    ...
+    ...
+    Heap Tracker lib: malloc(143360)... INFO: available heap (0 bytes)
+    Heap Tracker lib: _free(0xbedf5010, 143360)... INFO: available heap (143360 bytes)
+    Heap Tracker lib: malloc(144384)... WARNING: allocated heap (144384l bytes) is above available heap_limit (143360 bytes)
+    Heap Tracker lib: _free(0xbedf5010, 144384)... INFO: available heap (143360 bytes)
+    getFreeHeapMemory_malloc --> Currently free heap: 141Kb (144384 bytes)
+    ```
+
+- On failure, the OS's OOM Killer will SIGKILL the App's process. This will be the indicator that the global variable `heap_limit` in `heap_tracker_lib.c` should be lowered to a value below to the last successful allocation:
+
+    ```c
+    Remote debugging from host 192.168.35.1, port 61736
+    Starting Heap Tracker test application...
+    getFreeHeapMemory_realloc --> Heap status: max available(256000 bytes), allocated (0 bytes)
+    Heap-Tracker lib: _realloc(0,0,1024)=0xbee7b010... SUCCESS: available heap (254975 bytes)
+    Heap-Tracker lib: _realloc(0xbee7b010,1024,2048)=0xbee7b010... SUCCESS: available heap (253950 bytes)
+    Heap-Tracker lib: _realloc(0xbee7b010,2048,3072)=0xbeee9010... SUCCESS: available heap (252925 bytes)
+    ...
+    Heap-Tracker lib: _realloc(0xbeeff010,188416,189440)=0xbeeff010... SUCCESS: available heap (66375 bytes)
+    Heap-Tracker lib: _realloc(0xbeeff010,189440,190464)=0xbeeff010... SUCCESS: available heap (65350 bytes)
+    Heap-Tracker lib: _realloc(0xbeeff010,190464,191488)=0xbeeff010... SUCCESS: available heap (64325 bytes)
+
+    Child terminated with signal = 0x9 (SIGKILL)
+
+    ```
+
+## Key concepts
+The goal in using the Heap Tracker library, is to best-estimate the *actual* memory available to your application, in a more precise manner. This because the Linux OS (upon which the Azure Sphere OS is based) manages memory with a so called "[optimistic memory allocation strategy](https://man7.org/linux/man-pages/man3/malloc.3.html)", by means of which the OS almost never returns a `null` pointer upon memory allocation requests that would i.e. absolutely go beyond the device's availability.
+
+This is not an "overlooked" behavior in the OS, but rather a key feature for the OS to actually cope with the 100s/1000s of memory allocation requests that would never be satisfiable if immediately committed; this is achieved by accepting all requests and de facto "speculating" that statistically not all memory allocation requests will actually be committed immediately, if not committed at all.
```

Comment:
This statement should be reviewed by the OS team

Reply:
@tylerfox - can you confirm this?

Reply:
resolved offline